File: leetcode/cpp/P0859_BuddyStrings.cpp
#include <string>
#include <unordered_map>
#include <vector>

class Solution {
 public:
  bool buddyStrings(const std::string &A, const std::string &B) {
    if (A.length() != B.length()) {
      return false;
    }

    std::vector<int> distinct;

    for (int ii = 0; ii < A.length(); ++ii) {
      if (A[ii] != B[ii]) {
        distinct.push_back(ii);
        if (distinct.size() >= 3) {
          return false;
        }
      }
    }

    if (distinct.size() == 1) {
      return false;
    }

    if (distinct.size() == 0) {
      std::unordered_map<char, int> freq;
      for (char x : A) {
        ++freq[x];
        if (freq[x] >= 2) {
          return true;
        }
      }
      return false;
    }

    if (distinct.size() == 2) {
      return (A[distinct[0]] == B[distinct[1]] &&
              A[distinct[1]] == B[distinct[0]]);
    }

    return false;
  }
};

// TODO : Add main + basic test cases.

File: leetcode/python/P0859_BuddyStrings.py
from collections import Counter


class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        # strings have different length --> impossible --> false;
        # strings are equal --> need one char with frequency >= 2

        # we need to keep number of positions on which s and goal are different
        # if 2 distinct --> check if we can swap

        if len(s) != len(goal):
            return False

        distinct = []
        for idx in range(len(s)):
            if s[idx] != goal[idx]:
                distinct.append(idx)
                if len(distinct) > 2:
                    return False

        if len(distinct) == 1:
            return False
        elif len(distinct) == 2:
            first, second = distinct
            return s[first] == goal[second] and s[second] == goal[first]
        else:
            c = Counter(goal)
            for _, val in c.items():
                if val >= 2:
                    return True

        return False


# TODO add main + basic tests
